config/network: add ExternalIPs field

squeed opened the pull request:
This adds the ability to restrict external IPs. This is enforced by the openshift ExternalIPRanger admission controller.

squeed commented:
@danwinship PTAL

danwinship commented:
This admission controller belongs to @openshift/sig-network-edge doesn't it? Would it make more sense in some cluster-ingress-operator-related config? (Is there any other `ExternalIP` config anywhere? What about `LoadBalancer`?)

ironcladlou commented:
Our team hasn't assumed ownership of that admission controller, and we haven't had any discussions about trying to contextualize ExternalIP services in the ingress APIs.

@knobunc any thoughts here? It does seem like a gray area in that ExternalIP support conceptually could be framed as an ingress concern in terms of the security policy, but the actual implementation of ExternalIP support itself is SDN (e.g. iptables or nftables kernel routing).

squeed commented:
Aside: `Service.Spec.ExternalIP` is used for services of type `ClusterIP`. It's not a separate kind of service. It just tells kube-proxy to add an additional, manually-allocated service IP.

This doesn't really affect ingress, only kube-proxy and the admission controller (which lives in `pkg/network`, for what that's woth). It also has to do with blocks of IPs. I'd argue it's closer to "network's", technically.

From a user perspective, it is indeed a bit more grey. But since we already have the plumbing in the installer to generate the Network configuration, it makes sense to keep it here.

danwinship commented:
> This doesn't really affect ingress

right, but it's more edge-y than sdn-y

> only kube-proxy and the admission controller (which lives in `pkg/network`, for what that's worth).

That's not worth much since we're not the ones who put it there (eg, compare https://github.com/openshift/origin/pull/21126) or who wrote it originally.

But I'm fine with it being sig-network. I just wanted to double-check.

danwinship commented on config/v1/types_network.go:
```diff
@@ -24,7 +24,7 @@ type Network struct {
 // As a general rule, this SHOULD NOT be read directly. Instead, you should
 // consume the NetworkStatus, as it indicates the currently deployed configuration.
-// Currently, none of these fields may be changed after installation.
+// Currently, only the ExternalIPs can be updated
```
I think the "after installation" part is a useful clarification

danwinship commented on config/v1/types_network.go:
```diff
@@ -39,6 +39,12 @@ type NetworkSpec struct {
+	// If nil, any value is allowed for an ExternalIP. If the empty/zero
+	// policy is supplied, then ExternalIP is not allowed to be set.
+	// +optional
+	ExternalIPs *ExternalIPPolicy `json:"externalIPs,omitempty"`
```
Why is this being added to `config.openshift.io`'s NetworkSpec rather than `operator.openshift.io`'s? And even if it's going in the former, shouldn't it get copied to the latter too for consistency? (Otherwise it's the only piece of data that is in the `config` object but not the `operator` object.) And if it does go in the `operator` config, it seems like it would go in `spec.kubeProxyConfig`, in which case maybe there should be something parallel in the `config` object, in case we want to expose more proxy config later...

squeed replied:
(Answered Dan's question elsewhere, but bringing the answer here)

Because this isn't consumed by the network operator. Rather, it's observed by the cluster-kube-apiserver-operator. So, it needs to be cluster-level configuration.

squeed commented:
There are no CNO changes, since nothing it creates will need to consume this field. Instead, it will be part of the cluster-kube-apiserver-operator - https://github.com/openshift/cluster-kube-apiserver-operator/blob/master/pkg/operator/configobservation/network/observe_network.go

This config value only goes to the admission controller type in https://github.com/openshift/origin/blob/master/pkg/network/admission/apis/externalipranger/v1/types.go

danwinship commented on config/v1/types_network.go:
```diff
@@ -67,6 +73,17 @@ type ClusterNetworkEntry struct {
+	RejectedCIDRs []string `json:"rejectedCIDRs,omitempty"`
+
+	// allowedCIDRs is the list of allowed CIDRs.
+	AllowedCIDRs []string `json:"allowedCIDRs,omitempty"`
```
API-wise this is fine but it seems like `AllowCIDRs` should go first logically, and maybe not be `omitempty`?

/lgtm
other than this nitpick and the "after installation" nit above

squeed commented:
When implementing the controller (PR https://github.com/openshift/cluster-kube-apiserver-operator/pull/509), I realized we need to include IngressIPs as well for correctness. So, updated.

danwinship commented on config/v1/types_network.go:
```diff
@@ -39,6 +40,20 @@ type NetworkSpec struct {
+	// assigned an ExternalIP from this CIDR.
+	// Currently, only one value is supported
+	// +optional
+	IngressIPs []string `json:"ingressIPs,omitempty"`
```
um... is there *another* PR associated with this? Because what you're describing here is not something the admission controller does.

Also, could we rename this since it does not refer to `Ingress.networking.k8s.io`? (And maybe "`CIDRs`" rather than "`IPs`" too?)

squeed commented:
OK, changed this around quite a bit, and implemented the other controller necessary to configure this. It also has a much more reasonable naming scheme.

PRs that are part of this effort:
https://github.com/openshift/library-go/pull/465
https://github.com/openshift/cluster-openshift-controller-manager-operator/pull/104
https://github.com/openshift/cluster-kube-apiserver-operator/pull/509

This PR needs to merge first.


@danwinship @deads2k PTAL.

soltysh commented on config/v1/types_network.go:
```diff
@@ -67,6 +73,36 @@ type ClusterNetworkEntry struct {
+// field in a Service. If the zero struct is supplied, then none are permitted.
+type ExternalIPPolicy struct {
+	// allowedCIDRs is the list of allowed CIDRs.
+	// allowedCIDRs is the list of allowed CIDRs.
```
You have a duplicate comment.

soltysh commented on config/v1/types_network.go:
```diff
@@ -67,6 +73,36 @@ type ClusterNetworkEntry struct {
+	// Service.ExternalIP. These are assigned when the service is of type
+	// LoadBalancer. In general, this is only useful for bare-metal clusters.
+	// In Openshift 3.x, this was misleadingly called "IngressIPs".
+	// Currently, only one entry may be provided.
```
One entry and you pass array?

squeed replied:
We've learned that these sorts of things often get expanded. Right now, the controller only supports a single CIDR pool, but there's no technical reason that can't change. No sense in blocking ourselves in with an unnecessarily restrictive API.

danwinship commented on config/v1/types_network.go:
```diff
@@ -67,6 +73,36 @@ type ClusterNetworkEntry struct {
 
+// ExternalIPConfig specifies some IP blocks relevant for the ExternalIP field
+// of a Service resource.
+type ExternalIPConfig struct {
```
Need to explain on one of the two fields how they interact. (That is, that automatically-assigned egress IPs are allowed regardless of `policy`.)

squeed replied:
very good point.

squeed commented:
Thanks for the review, updated.

danwinship commented:
/lgtm
